### PR TITLE
Beta: more async infrastructure

### PR DESCRIPTION
**Pull request by richardm-stripe**
## Summary
Adds async mirrors for more "infrastructure" methods.
Some of these I exercise directly in tests here, the rest are exercised in #1171.

### CODE REVIEW

File: stripe/_api_resource.py [hunk 1]
```diff
@@ -137,24 +183,62 @@ def _static_request(
...
             base_address=base_address,
             api_mode=api_mode,
         )
+
+    @classmethod
+    def _static_request_stream(
+        cls,
+        method,
+        url,
+        params=None,
```
Comment (anniel-stripe) [inline stripe/_api_resource.py:211]: nit
```suggestion
        params: Optional[Mapping[str, Any]] = None,
```

File: stripe/_api_resource.py [hunk 2]
```diff
@@ -137,24 +183,62 @@ def _static_request(
...
+            base_address=base_address,
+            api_mode=api_mode,
+        )
+
+    @classmethod
+    async def _static_request_stream_async(
+        cls,
+        method,
+        url,
+        params=None,
```
Comment (anniel-stripe) [inline stripe/_api_resource.py:231]:
```suggestion
        params: Optional[Mapping[str, Any]] = None,
```

File: stripe/_list_object.py [hunk 3]
```diff
@@ -160,18 +192,30 @@ def next_page(self, **params: Unpack[RequestOptions]) -> Self:
...
+    async def next_page_async(self, **params: Unpack[RequestOptions]) -> Self:
         if not self.has_more:
             request_options, _ = extract_options_from_dict(params)
             return self._empty_list(
                 **request_options,
             )
 
+        return await self.list_async(**self._get_filters_for_next_page(params))
+
+    def _get_filters_for_previous_page(
```
Comment (anniel-stripe) [inline stripe/_list_object.py:216]: Nice!

File: stripe/_search_result_object.py [hunk 4]
```diff
@@ -54,6 +55,23 @@ def search(self, **params: Mapping[str, Any]) -> Self:
             ),
         )
 
+    async def _search_async(self, **params: Mapping[str, Any]) -> Self:
+        url = self.get("url")
+        if not isinstance(url, str):
+            raise ValueError(
+                'Cannot call .list on a list object without a string "url" property'
+            )
```
Comment (anniel-stripe) [inline stripe/_search_result_object.py:63]: Could we move this into a `_get_url_for_search`?

File: stripe/_search_result_object.py [hunk 5]
```diff
@@ -121,3 +150,22 @@ def next_search_result_page(
...
+                stripe_account=options.get("stripe_account"),
+            )
+
+        params_with_filters = dict(self._retrieve_params)
+        params_with_filters.update({"page": self.next_page})
+        params_with_filters.update(params)
+
+        return await self._search_async(
+            **params_with_filters,
+        )
```
Comment (anniel-stripe) [inline stripe/_search_result_object.py:171]: Same here, move into a `_get_filters_for_next_page` method

**Review by anniel-stripe — APPROVED** [review]

File: stripe/_search_result_object.py (hunk 4 above)
Comment (richardm-stripe) [reply stripe/_search_result_object.py]: Done in [_get_url_for_search and _maybe_empty_result](https://github.com/stripe/stripe-python/pull/1219/commits/d38887b367d0b0cd8c63aef60debd6c86840e4d1)

File: stripe/_search_result_object.py (hunk 5 above)
Comment (richardm-stripe) [reply stripe/_search_result_object.py]: Done in [_get_url_for_search and _maybe_empty_result](https://github.com/stripe/stripe-python/pull/1219/commits/d38887b367d0b0cd8c63aef60debd6c86840e4d1)

**Pull request merged**